TASK-71: DM's Backend

bit172 opened the pull request:
This PR closes #247.

**Steps to Test**

1. Open your favorite GraphQL playground
2. Make sure you have users in your database
2. Use the createDm/EventChannel mutation & getDm/EventChannel query
    - getDmChannel should return null if the users do not have a channel already
3. Check if posts behave the same as before (creating, subscribing, querying)
    - The placeholder subscriber is present on the announcement page

RGPosadas committed:
feat: use single table inheritance for EventChannel and DirectMessageChannel

- Using single table inheritance enables us to customize channels for DMs as they have different needs compared to event channels

Refs #247

Co-authored-by: Jonathan <bit172@gmail.com>

RGPosadas committed:
feat(channel.service): add findUsersInDmChannel function to search for participating users in a DM channel

- Add eventChannel and dmChannel repositories as well as their respective get/create ops

Refs #247

Co-authored-by: Jonathan <bit172@gmail.com>

RGPosadas committed:
feat(channel.resolver): add getDmChannel and getEventChannel queries

Refs #247

Co-authored-by: Jonathan <bit172@gmail.com>

bit172 committed:
feat(channel.resolver): add resolvers for new channels

Ref: #247

Co-authored-by: Ritchelle Posadas  <ritchelle.posadas@gmail.com>

bit172 committed:
test(channel): add tests for service/resolver

Ref: #247

Co-authored-by: Ritchelle Posadas  <ritchelle.posadas@gmail.com>

cindyslittleplanet reviewed (approved):
LTGM! 🚀 Good job! 


**Subtasks**
- [x] 71.1 Single table inheritance for DirectMessageChannel and EventChannel
- [x] 71.2 Users are related to channels

**Acceptance Criteria**
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every new component and function
- [x] The application should still work with seeded data

osahtout commented on apps/mull-api/src/app/channel/inputs/channel.input.ts:
```diff
@@ -8,10 +10,17 @@ export class ChannelInput implements Partial<Channel> {
 }
+
+@InputType()
+export class CreateDmChannelInput implements Partial<DirectMessageChannel> {
```
I know this may seem like I'm bias but I would suggest either have DM everywhere so `CreateDmChannelInput` and `DmChannel` and write somewhere that DM stands for direct messages or have have the full name everywhere: `CreateDirectMessageChannelInput ` and `DirectMessageChannel`

osahtout commented on apps/mull-api/src/app/channel/channel.service.ts:
```diff
@@ -1,26 +1,59 @@
     return true;
   }
+
+  async findUsersInDmChannel(userId1: number, userId2: number): Promise<DirectMessageChannel> {
```
The name is a little confusing. I thought this returns the users of a dm channel but it seems like it returns the channel given the users so it would be something like findDMChannelByUsersIds.

RGPosadas replied:
Done

osahtout commented on apps/mull-api/src/app/channel/channel.service.ts:
```diff
@@ -1,26 +1,59 @@
+      for (const participant of participants) {
+        if (participant.id === userId1 || participant.id === userId2) count++;
+      }
+      return count === 2;
```
This isn't a blocker just maybe a suggestion if we have time to refactor this. 

Can we have a column that concatenates the users ids to find the channel with a separator in the middle?  
Say we have user with id 1 and user with id 2, then we add to the database the string `1+2`. say we have a user with id 111 and user with id 11 then we would have the string `11+111`. we would always have the smallest number first .

Maybe it's a bad idea I'm not sure. Just wanted to hear your opinions

RGPosadas replied:
After talking, we decided not to do this.

leojrsilao reviewed (approved):
LGTM!! 🚀 

Linters and unit tests pass.
DM Mutations are functional and users are still related to channels as intended.

r-saba commented on apps/mull-api/src/app/channel/channel.resolver.ts:
```diff
@@ -1,20 +1,27 @@
-  async createChannel(
-    @Args('input') input: CreateChannelInput,
+  async createEventChannel(
+    @Args('input') input: CreateEventChannelInput,
     @Args('eventId', { type: /* istanbul ignore next */ () => Int }) eventId?: number
```
`eventId` is mandatory now, we should remove the second optional argument here and have `eventId` as part of `CreateEventChannelInput`

RGPosadas replied:
Done

RGPosadas replied:
Upon further inspection, createEventChannel does not need to exposed in their respective service/resolver files as createEvent is the one creating those channels through `cascade: true`. Therefore, we will be removing these functions as they are not used anywhere else.

r-saba commented on apps/mull-api/src/app/channel/channel.resolver.ts:
```diff
@@ -1,20 +1,27 @@
+    return this.channelService.createEventChannel(input, eventId);
+  }
+
+  @Mutation(/* istanbul ignore next */ () => DirectMessageChannel)
```
Noticed here we return a Channel but in `createEventChannel` we return a Boolean. Should we make it so both return a channel or both return a boolean to be consistent?

bit172 replied:
Sure, we will return it for all creates/deletes in case we need it in the future

r-saba commented on apps/mull-api/src/app/channel/inputs/channel.input.ts:
```diff
@@ -8,10 +10,17 @@ export class ChannelInput implements Partial<Channel> {
+@InputType()
+export class CreateDmChannelInput implements Partial<DirectMessageChannel> {
+  @Field(() => [UserInput])
+  @ArrayMaxSize(2)
```
We should also at add a Min here. I can create a DM channel with 0 or 1 participant

r-saba commented on apps/mull-api/src/app/channel/channel.resolver.ts:
```diff
@@ -28,4 +35,21 @@ export class ChannelResolver {
   async getChannel(@Args('id', { type: /* istanbul ignore next */ () => Int }) id: number) {
     return this.channelService.getChannel(id);
   }
```
Do you think we still need this or is it just unnecessary code now? Lets continue this discussion here https://github.com/RGPosadas/Mull/pull/250#discussion_r588926946

r-saba commented on apps/mull-api/src/app/channel/channel.service.spec.ts:
```diff
@@ -1,50 +1,128 @@
-  it('should get a channel', async () => {
-    const getChannel = await service.getChannel(mockAllChannels[0].id);
-    expect(getChannel).toBe(mockAllChannels[0]);
+  it('should get a generic channel', async () => {
+    const returnedChannel = await service.getChannel(mockAllChannels[0].id);
+    expect(returnedChannel).toBe(mockAllChannels[0]);
+  });
```
The way I was thinking about `channel`, is that it would be abstract class and we'd have other channels that implement the interface. Feel like this would be a clean and OO approach. What do you think, do we ever really want to access a channel that doesn't implement the generic one?

bit172 replied:
We will delete this resolver.

r-saba commented on apps/mull-api/src/app/channel/channel.resolver.ts:
```diff
@@ -28,4 +35,21 @@ export class ChannelResolver {
+
+  @Query(/* istanbul ignore next */ () => DirectMessageChannel, { nullable: true })
+  @UseGuards(AuthGuard)
+  async getDmChannel(
+    @Args('participants', { type: /* istanbul ignore next */ () => [UserInput] })
+    participants: UserInput[]
+  ) {
+    const [usr1, usr2] = participants;
+    const dmChannel = await this.channelService.findUsersInDmChannel(usr1.id, usr2.id);
+    return dmChannel ? dmChannel : null;
```
This isn't secure. Since it's using AuthGuard, any user can query an DM channel by passing userId 1 and UserId 2.
We should use `AuthenticatedUser` for userId 1

r-saba replied:
We want to make sure security is handled for all DM channel queries/mutations

bit172 replied:
Thank you for bringing this up, we will check for security in our resolvers

r-saba commented on apps/mull-api/src/app/channel/channel.resolver.ts:
```diff
@@ -28,4 +35,21 @@ export class ChannelResolver {
+
+  @Query(/* istanbul ignore next */ () => EventChannel)
+  @UseGuards(AuthGuard)
+  async getEventChannel(@Args('id', { type: /* istanbul ignore next */ () => Int }) id: number) {
+    return this.channelService.getEventChannel(id);
+  }
```
How do we want to handle security for event channels? In #251 I implemented the logic already so it would be redundant for you to do it again here

RGPosadas replied:
Done!

r-saba reviewed (changes requested):
Blocking because we should handle security of mutations/queries in this PR too

bit172 committed:
refactor(channel): move validation function into entity class
- Implement strategy pattern for validation functions of Channels

Ref: #247

Co-authored-by: Ritchelle Grace Posadas <RGPosadas@users.noreply.github.com>
Co-authored-by: Ragith Saba <ragith.coolkid@gmail.com>

RGPosadas committed:
test(channel): update tests to include validation

- Increased security of resolvers by using AuthenticatedUser instead of using an array of participants
- Update frontend to use the renamed query
- Removed CreateDmChannelInput in favor of AuthenticatedUser

Refs #247

Co-authored-by: Jonathan Hsu <bit172@gmail.com>

RGPosadas committed:
test(post): update tests to include validation

Refs #247

Co-authored-by: Jonathan Hsu <bit172@gmail.com>

RGPosadas committed:
feat(channel): removed createEventChannel service and resolver methods

We removed these methods as createEvent is the one cthat is creating these event channels, therefore there is currently no need to have these methods exposed as it it not being used functionally.

Refs #247

Co-authored-by: Jonathan Hsu <bit172@gmail.com>

RGPosadas committed:
chore: rename all instances of dm to DirectMessage

Refs #247

Co-authored-by: Jonathan Hsu <bit172@gmail.com>

RGPosadas committed:
chore: remove unused repository

Co-authored-by: Jonathan Hsu <bit172@gmail.com>

r-saba commented on apps/mull-api/src/app/channel/channel.resolver.spec.ts:
```diff
@@ -4,39 +4,36 @@ import { MockType } from '../user/user.service.spec';
     ...mockEventChannelData,
   })),
-  createDmChannel: jest.fn((mockChannelData: CreateDmChannelInput) => ({ ...mockChannelData })),
+  createDmChannel: jest.fn().mockReturnValue(mockAllDirectMessageChannels[0]),
```
Can we implement the logic instead of hard coding the return value? We've done this in other tests and it's made the debugging process/reusability of these mock implementations difficult. If it doesn't take too much time.

r-saba reviewed (changes requested):
Noticed I can create a multiple DM channels with the same list of participants. This would be an issue when trying to retrieve a DM channel between two users.

RGPosadas committed:
feat(dmChannel): update createDirectMessageChannel to handle case where there already exists a DM channel between two users

Refs #247

Co-authored-by: Jonathan Hsu <bit172@gmail.com>

r-saba commented on apps/mull-api/src/app/channel/channel.service.spec.ts:
```diff
@@ -1,66 +1,148 @@
+    });
+    expect(
+      async () => await service.createDirectMessageChannel(mockAllUsers[0].id, mockAllUsers[1].id)
+    ).rejects.toThrow(UnauthorizedException);
```
Some error messages resulted from this test. To fix before merging
```suggestion
    ).rejects.toThrow(ConflictException);
```

r-saba reviewed (approved):
LGTM, left a small comment to fix before merging
**Acceptance Criteria**
- [X] Documentation present where necessary
- [X] Unit and UI tests are implemented for every new component and function
- [X] The application should still work with seeded data
